src/ring_theory/adjoin/basic.lean

```diff
@@ -110,6 +117,72 @@ lemma adjoint_prod_le (s : set A) (t : set B) :
   adjoin R (set.prod s t) ≤ (adjoin R s).prod (adjoin R t) :=
 adjoin_le $ set.prod_mono subset_adjoin subset_adjoin
 
+lemma adjoin_le_prod (s) (t) : adjoin R ((set.insert (1, 0) ((linear_map.inl R A B) '' s)) ∪
+  (set.insert (0, 1) ((linear_map.inr R A B) '' t))) ≤ (adjoin R s).prod (adjoin R t) :=
+begin
+  refine adjoin_le (set.union_subset
+    (set.insert_subset.2 ⟨set.mem_prod.2 ⟨set_like.mem_coe.2 $ one_mem _,
+    set_like.mem_coe.2 $ zero_mem _⟩, λ x, _⟩)
+    (set.insert_subset.2 ⟨set.mem_prod.2 ⟨set_like.mem_coe.2 $ zero_mem _,
+    set_like.mem_coe.2 $ one_mem _⟩, λ x, _⟩)),
+  { simp_intros hx,
+    obtain ⟨X, hXmem, hXx⟩ := hx,
+    exact ⟨subset_adjoin (by rwa ←hXx), by { rwa ←hXx, exact zero_mem _ }⟩ },
+  { simp_intros hx,
+    obtain ⟨X, hXmem, hXx⟩ := hx,
+    exact ⟨by { rwa ←hXx, exact zero_mem _ }, subset_adjoin (by rwa ←hXx)⟩ }
+end
+
+lemma adjoint_prod_fst_mem (B) [semiring B] [algebra R B] {s} {x : A} (h : x ∈ adjoin R s) :
+  ((x, 0) : (A × B)) ∈ (adjoin R (set.insert (1, 0) ((linear_map.inl R A B) '' s))) :=
+begin
+  let S := adjoin R (set.insert (1, 0) ((linear_map.inl R A B) '' s)),
+  refine @adjoin_induction R A _ _ _ _
+    (λ a, ((a, 0) : (A × B)) ∈ (adjoin R (set.insert (1, 0) ((linear_map.inl R A B) '' s)))) x h
+    (λ a ha, subset_adjoin $ set.subset_insert _ _ ⟨a, ⟨ha, rfl⟩⟩)
+    (λ r, _)
+    (λ y z hy hz, by simpa [hy, hz] using subalgebra.add_mem _ hy hz)
+    (λ y z hy hz, by simpa [hy, hz] using subalgebra.mul_mem _ hy hz),
+  have : ((1, 0) : A × B) ∈ S := subset_adjoin (set.mem_insert _ _),
+  replace this := subalgebra.smul_mem S this r,
+  rw [prod.smul_mk, smul_zero] at this,
+  convert this,
+  exact algebra_map_eq_smul_one r,
+end
+
+lemma adjoint_prod_snd_mem (A) [semiring A] [algebra R A] {t} {x : B} (h : x ∈ adjoin R t) :
+  ((0, x) : (A × B)) ∈ (adjoin R (set.insert (0, 1) ((linear_map.inr R A B) '' t))) :=
+begin
+  let T := adjoin R (set.insert (0, 1) ((linear_map.inr R A B) '' t)),
+  refine @adjoin_induction R B _ _ _ _
+    (λ b, ((0, b) : (A × B)) ∈ (adjoin R (set.insert (0, 1) ((linear_map.inr R A B) '' t)))) x h
+    (λ b hb, subset_adjoin $ set.subset_insert _ _ ⟨b, ⟨hb, rfl⟩⟩)
+    (λ r, _)
+    (λ y z hy hz, by simpa [hy, hz] using subalgebra.add_mem _ hy hz)
+    (λ y z hy hz, by simpa [hy, hz] using subalgebra.mul_mem _ hy hz),
+  have : ((0, 1) : A × B) ∈ T := subset_adjoin (set.mem_insert _ _),
+  replace this := subalgebra.smul_mem T this r,
+  rw [prod.smul_mk, smul_zero] at this,
+  convert this,
+  exact algebra_map_eq_smul_one r,
+end
+
+lemma adjoin_eq_prod (s) (t) : adjoin R ((set.insert (1, 0) ((linear_map.inl R A B) '' s)) ∪
+  (set.insert (0, 1) ((linear_map.inr R A B) '' t))) = (adjoin R s).prod (adjoin R t) :=
```

Comment:
What are your thoughts on this version of the statement? It's true, right?
```
lemma adjoin_eq_prod (s : set A) (t : set B) :
  adjoin R (s.prod t) ⊔ adjoin R ({(1, 0), (0, 1)}) = (adjoin R s).prod (adjoin R t) :=
```

Reply:
This lemma would certainly help either prove that or use that definition:
```
lemma adjoin_union (s t : set A) :
  adjoin R (s ∪ t) = adjoin R s ⊔ adjoin R t :=
show adjoin R (s ⊔ t) = _,
from galois_connection.l_sup (algebra.gc : galois_connection _ coe)
```

Reply:
```lean
lemma adjoin_prod_adjoin (s : set A) (t : set B) :
  (adjoin R s).prod (adjoin R t) = adjoin R (s.prod t) ⊔ adjoin R ({(1, 0), (0, 1)}) :=
```
Is true I think. But even better I think we also have
```lean
lemma adjoin_prod_adjoin' (s : set A) (t : set B) :
  (adjoin R s).prod (adjoin R t) = adjoin R ((s.prod t) ∪  {(1, 0), (0, 1)}) :=
```
(This is better since `(s.prod t) ∪  {(1, 0), (0, 1)}` is finite if `s` and `t` are, so perfect for finitely generated algebras.). I will modify the proof to use this.

Reply:
Ah, this is false! If `s = A` and `t =∅`, then `(adjoin R s).prod (adjoin R t) = A × R`, but `s.prod t = ∅`, so `adjoin R ((s.prod t) ∪  {(1, 0), (0, 1)}) = R × R`!

Reply:
Nice find. Is this one true? (and is it useful?)
```
lemma adjoin_prod_adjoin' (s : set A) (t : set B) :
  (adjoin R s).prod (adjoin R t) = adjoin R ((s ∪ {0, 1}).prod (t ∪ {0, 1})):=
```

Reply:
Take `R=ℝ`, `A=B=ℝ[X]` and `s = t = ℤ[X]`. We have `adjoin R s = adjoin R t = ℝ[X]`, so `(adjoin R s).prod (adjoin R t) = ℝ[X] × ℝ[X]`. On the other hand `s ∪ {0, 1} = t ∪ {0, 1} = ℤ[X]`, so `(s ∪ {0, 1}).prod (t ∪ {0, 1}) = ℤ[X] × ℤ[X]` and `adjoin R ((s ∪ {0, 1}).prod (t ∪ {0, 1})) = adjoin ℝ (ℤ[X] × ℤ[X] )`.

Now, `(X, πX) ∈ ℝ[X] × ℝ[X]` but `(X, πX) ∉ adjoin ℝ (ℤ[X] × ℤ[X] )`: indeed `ℤ[X] × ℤ[X]` is already a ring, so the only thing you can do to enlarge it to `ℝ (ℤ[X] × ℤ[X] )` is multiplying by a real number, but if `(P, Q) ∈ ℤ[X] × ℤ[X] ` and `α ∈ ℝ` are such that `α(P, Q) = (X, πX)` then `αQ = πX` and so `α = π` (I took `π` but indeed any irrational works) and `πP = X` is absurd.

Reply:
I think that in general is very difficult to write `(adjoin R s).prod (adjoin R t)` as `adjoin R (?.prod ?)`: if you write a general element of the RHS as linear combination of products of elements of `?.prod ?` you see that you have much less freedom than in doing the same thing for `(adjoin R s).prod (adjoin R t)`, where you can mix the coefficients as you wish.